caiman/source_extraction/cnmf/lazy_arrays.py

```diff
@@ -0,0 +1,444 @@
+
+    def __getitem__(
+            self,
+            item: Union[int, Tuple[slice_or_int_or_range]]
+    ):
+        if isinstance(item, int):
+            indexer = item
+
+        # numpy int scaler
+        elif isinstance(item, np.integer):
+            indexer = item.item()
+
+        # treat slice and range the same
+        elif isinstance(item, (slice, range)):
+            indexer = item
+
+        elif isinstance(item, tuple):
+            if len(item) > len(self.shape):
+                raise IndexError(
+                    f"Cannot index more dimensions than exist in the array. "
+                    f"You have tried to index with <{len(item)}> dimensions, "
+                    f"only <{len(self.shape)}> dimensions exist in the array"
+                )
+
+            indexer = item[0]
+
+        else:
+            raise IndexError(
+                f"You can index LazyArrays only using slice, range, int, or tuple of slice and int, "
+                f"you have passed a: <{type(item)}>"
+            )
+
+        # treat slice and range the same
+        if isinstance(indexer, (slice, range)):
+            start = indexer.start
+            stop = indexer.stop
+            step = indexer.step
+
+            if start is not None:
+                if start > self.n_frames:
+                    raise IndexError(f"Cannot index beyond `n_frames`.\n"
+                                     f"Desired frame start index of <{start}> "
+                                     f"lies beyond `n_frames` <{self.n_frames}>")
+            if stop is not None:
+                if stop > self.n_frames:
+                    raise IndexError(f"Cannot index beyond `n_frames`.\n"
+                                     f"Desired frame stop index of <{stop}> "
+                                     f"lies beyond `n_frames` <{self.n_frames}>")
+
+            if step is None:
+                step = 1
+
+            # convert indexer to slice if it was a range, allows things like decord.VideoReader slicing
+            indexer = slice(start, stop, step)  # in case it was a range object
+
+            # dimension_0 is always time
+            frames = self._compute_at_indices(indexer)
+
+            # index the remaining dims after lazy computing the frame(s)
+            if isinstance(item, tuple):
+                if len(item) == 2:
+                    return frames[:, item[1]]
+                elif len(item) == 3:
+                    return frames[:, item[1], item[2]]
+
+            else:
+                return frames
+
+        elif isinstance(indexer, int):
+            return self._compute_at_indices(indexer)
+
+    def __repr__(self):
+        return f"{self.__class__.__name__} @{hex(id(self))}\n" \
+               f"{self.__class__.__doc__}\n" \
+               f"Frames are computed only upon indexing\n" \
+               f"shape [frames, x, y]: {self.shape}\n"
+
+
+class LazyArrayRCM(LazyArray):
+    """LazyArray for reconstructed movie, i.e. A ⊗ C"""
```

Comment:
Maybe just AC  (that ⊗ symbol is usually for outer product)

Reply:
But it is outer product?

Reply:
Vector for spatial and vector for temporal are multiplied to product a matrix if I understand it correctly: https://en.wikipedia.org/wiki/Outer_product#Definition

Reply:
The key trick is `self.spatial @ self.temporal[:, indices]`, iiic, and this is basically `AC_`, or matrix multiplication on limited number of time points (columns) of C.  So I don't think it's using outer products explicitly. I guess it's outer product in the broad sense that all matrix multiplication is a sum of outer products :smile: , but it's basically just matrix multiplication limited to certain columns of the second matrix. Similarly for background `bf` it will be the same (except limited columns of bg fluctuations matrix `f`).

Reply:
Isn't it outer product because you're doing vector x  vector = matrix? Andrea always referred to it as the outer product too.

Reply:
I generally thought:

vector.T * vector = single number, therefore inner product
matrix * matrix = matrix, two transformation so you just get another matrix
vector * vector.T = matrix, outer product

Reply:
maybe we can email Gibert Strang :joy:

Reply:
But this is matrix x matrix, generally, right?

Here we have `A` (spatial component matrix in `self.spatial`) times some truncated `C` in `self.temporal[:, indices]` (indices picked based on how lazy you want to be). It's either matrix times vector (if you are picking just one index) or matrix time matrix (if you pick a block of indices). Not sure what I'm missing here it's just AC (but C is not all of C, because we're only picking certain columns).

Where we talk about outer product is when building up one component of the full neural model: a_i ⊗ c_i (one column from A and one row of C) and that gives you the single neuron movie.  But this is different, this is `AC_` (where I am using C_ to refer to the C matrix block limited to `indices` columns. Unless I'm missing something?